rfcs/0107-version-normalization.md

```diff
@@ -0,0 +1,208 @@
+    `unstable`, `trunk`, `experimental`, `staging`, `X.Y` (where `X` and
+    `Y` are numbers) etc.
+
+## Design
+
+- For a labeled snapshot:
+  - `version` should be constituted of the version of the snapshot, as defined
+    by the program team, without any alphabetical characters (e.g. "v", "rel")
+    prepending it.
+    - Alphabetical characters following the first numerical character of version
+      (as defined above) should be maintained, except optionally those clearly
+      used as separators, in which case they are replaced by dots (emulating a
+      typical dot-separated version).
+
+- For an unlabeled snapshot:
+  - `version` should be constituted of a concatenation of the elements below in
+    this order:
+      - the version of the latest previous labeled snapshot (on the same branch,
+        when applicable), according to the rules defined before for labeled
+        snapshots;
+        - If the project never released a labeled snapshot, `0.pre` should be
+          used as default.
+     - the string `+unstable=YYYY-MM-DD`, where `YYYY-MM-DD` denotes the date
+       the mentioned unlabeled snapshot was distributed.
+
+- Some atypical programs have special considerations:
+  - Linux kernel modules:
+    - Besides the rules established above for typical snapshots (whether labeled
+      or unlabeled), `version` shoud have appended `+linux=XX.XX.XX`, where
+      `XX.XX.XX` is the corresponding Linux kernel version used to build the
+      aforementioned module.
+
+# Examples and Interactions
+[examples-and-interactions]: #examples-and-interactions
+
+Some useful examples:
+
+- Bochs is a typical Sourceforge-hosted project; its labeled snapshots can be
+  fetched from tarballs obtained via URLs like
+  <https://sourceforge.net/projects/bochs/files/bochs/2.6.11/>
+
+  For this example, we have `pname = "bochs"; version = "2.6.11";`.
+
+- MPV is a typical Github-hosted program; its labeled snapshots can be fetched
+  from tarballs obtained via URLs like
+  <https://github.com/mpv-player/mpv/releases/tag/v0.33.1>.
+
+  For this example, we get rid of the `"v"` prepended to the version tag: `pname
+  = "mpv"; version = "0.33.1";`.
+
+- SDL2 is hosted on Github; its latest labeled version can be downloaded from
+  <https://github.com/libsdl-org/SDL/releases/tag/release-2.0.14>. Therefore we
+  have `pname = "SDL2"; version = "2.0.14";`.
+
+  _However_, this labeled version was released December 21, 2020, while the
+  latest change was done in May 28, 2021.
+
+  Therefore, for this particular unlabeled releases of SDL2, we have `pname =
+  "SDL2"; version = "2.0.14+unstable=2021-05-28";`.
+
+- Cardboard is a typical Gitlab-hosted program. It has no labeled release yet,
+  therefore we use `0.pre` as default dummy stable version; further, the latest
+  commit was made on May 10, 2021.
+
+  Therefore, we have `pname = "cardboard"; version =
+  "0.pre+unstable=2021-05-21";`.
```

Comment:
```suggestion
  Therefore, we have `pname = "cardboard"; version =
  "0.pre+unstable=2021-05-21";`.
```
I really do not like 0.pre and then the additional unstable part. Any form of unstable already implies a pre-release, especially if it's in a completely unreleased program.

Reply:
This is a semantical quirk. This default string for unreleased releases should satisfy two constraints:

- Start with a digit (because `parseDrvName` splits it on the first `-` before a number);
- Be the smallest possible, in the sense of `compareVersions`

Therefore, a `version="unstable-DATE";` can't be, because it will complain with `parseDrvName` (as already said in the RFC motivation).

The string  `0.0.0` was found in the wild. Therefore, it makes that "special string" essentially blocked.

The next idea was to use `0.pre` because `pre` has the special property of being the smallest thing possible.

The `unstable` name was the best we could think until now, and it eases the transition.

Reply:
See also https://github.com/NixOS/rfcs/pull/107#discussion_r716214480 and the following discussion.